### PR TITLE
Fix #5718: Solana Account Balances

### LINKED ISSUE #5718: Fetch Solana balance
### Description: 

Currently we are only correctly fetch balance for Ethereum accounts. Need to update wallet service helpers to fetch Solana balance, and fetch Solana SPL token balances correctly.

### PR DESCRIPTION
**Pull request by StephenHeaps**
## Summary of Changes
- Update `BraveWalletJsonRpcService` extensions to support fetching Solana SOL balances and SPL Token balances.
- Add new `BraveWalletAssetRatioService` extension to fallback to fetching individual prices when the group fetch fails. The group fetch will fail if the price call fails for _just one_ of the token symbols. This is possible when you have a custom minted SPL Token.
- Fix adding a custom asset for Solana network. Currently it will add to the selected network. New issue created: #5725
- Fix account details view using currently selected network instead of account's coin type network. This caused Solana assets to show in Ethereum accounts when a Solana network was selected (and vice versa).

This pull request fixes #5718

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New or updated UI has been tested across:
  - [x] Light & dark mode
  - [x] Different size classes (iPhone, landscape, iPad)
  - [x] Different dynamic type sizes

## Test Plan:
1. Switch to a Solana account with some balance on SOL and SPL tokens.
2. Verify balance updates correctly in Portfolio
3. Tap the SOL asset in Portfolio.
4. Verify balance displays correctly in Asset Details
5. Tap the SPL Token asset in Portfolio
6. Verify balance displays correctly in Asset Details
7. Tap back, and switch to Accounts tab
8. Select a Solana account
9. Verify balance displays correctly in Account Details
10. Switch back to an Ethereum account with some balance
10. Verify balance still displays correctly / as expected.

## Screenshots:

 |  |  | 
--|--|--|--

## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue and pull request is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: BraveWallet/Extensions/RpcServiceExtensions.swift [hunk 1]
```diff
@@ -83,42 +117,109 @@ extension BraveWalletJsonRpcService {
...
+    } else if token.isErc721 {
+      erc721TokenBalance(
+        token.contractAddress,
+        tokenId: token.tokenId,
+        accountAddress: accountAddress,
+        chainId: network.chainId,
+        completion: completion
+      )
+    } else {
+      let errorMessage = "Unable to fetch ethereum balance for \(token.symbol) token in account address '\(accountAddress)'"
```
Comment (StephenHeaps) [inline BraveWallet/Extensions/RpcServiceExtensions.swift:221]: Currently, this error message will not see the user. This function is private to this file, and when status is not `success` we return nil value in the completion blocks on [L30](https://github.com/brave/brave-ios/pull/5726/files#diff-318bfcb7b1ee7544b84a0e90c73b12278955ea7b4e1f462f6ad5aa93c49ec33cR30) & [L129](https://github.com/brave/brave-ios/pull/5726/files#diff-318bfcb7b1ee7544b84a0e90c73b12278955ea7b4e1f462f6ad5aa93c49ec33cR129) of this file

File: BraveWallet/Crypto/Stores/PortfolioStore.swift [hunk 2]
```diff
@@ -121,14 +121,12 @@ public class PortfolioStore: ObservableObject {
 
   /// Fetches the prices for a given list of symbols, giving a dictionary with the price for each symbol
   private func fetchPrices(for symbols: [String], completion: @escaping ([String: String]) -> Void) {
-    assetRatioService.price(
+    assetRatioService.priceWithIndividualRetry(
```
Comment (nuo-xu) [inline BraveWallet/Crypto/Stores/PortfolioStore.swift:124]: this is good. iirc we currently behaves differently compared to desktop and android.

File: BraveWallet/Extensions/RpcServiceExtensions.swift [hunk 3]
```diff
@@ -83,42 +117,109 @@ extension BraveWalletJsonRpcService {
...
+              completion(nil)
+            }
+          }
+        } else {
+          splTokenAccountBalance(
+            accountAddress,
+            tokenMintAddress: token.contractAddress,
+            chainId: network.chainId
+          ) { amount, lamports, uiAmountString, status, errorMessage in
+            guard status == .success else {
```
Comment (nuo-xu) [inline BraveWallet/Extensions/RpcServiceExtensions.swift:168]: what are `lamports` and `uiAmountString` do here?

File: BraveWallet/Extensions/RpcServiceExtensions.swift [hunk 4]
```diff
@@ -15,42 +15,76 @@ extension BraveWalletJsonRpcService {
...
-      } else {
+      case .sol:
+        if token.symbol == network.nativeToken.symbol {
+          solanaBalance(account.address, chainId: network.chainId) { lamports, status, errorMessage in
+            guard status == .success else {
+              completion(nil)
+              return
+            }
+            let formatter = WeiFormatter(decimalFormatStyle: decimalFormatStyle)
+            if let valueString = formatter.decimalString(
```
Comment (nuo-xu) [inline BraveWallet/Extensions/RpcServiceExtensions.swift:53]: hmmm probably rename this formatter or at least update the method docs `value` is now a wei string in ethererum or a lamports string in solana

File: BraveWallet/Extensions/RpcServiceExtensions.swift (hunk 3 above)
Comment (StephenHeaps) [reply BraveWallet/Extensions/RpcServiceExtensions.swift:168]: These are the names from the callback definition, they are different precision of the amount, ex. amount=998999100000000000, lamports=9, uiAmountString=998999100

File: BraveWallet/Extensions/RpcServiceExtensions.swift (hunk 4 above)
Comment (StephenHeaps) [reply BraveWallet/Extensions/RpcServiceExtensions.swift:53]: Updated the function doc in [18e3530](https://github.com/brave/brave-ios/pull/5726/commits/18e35301424c8ecaa8493a0c92b6c979fc2572d1)

**Review by nuo-xu — APPROVED** [review]

**Review by kylehickinson — APPROVED** [review]

**Pull request merged**